src/csharp/Grpc.Core/Channel.cs

```diff
@@ -130,27 +130,33 @@ public ChannelState State
         // cached handler for watch connectivity state
         static readonly BatchCompletionDelegate WatchConnectivityStateHandler = (success, ctx, state) =>
         {
-            var tcs = (TaskCompletionSource<object>) state;
-            if (success)
-            {
-                tcs.SetResult(null);
-            }
-            else
-            {
-                tcs.SetCanceled();
-            }
+            var tcs = (TaskCompletionSource<bool>) state;
+            tcs.SetResult(success);
         };
 
         /// <summary>
         /// Returned tasks completes once channel state has become different from 
         /// given lastObservedState. 
         /// If deadline is reached or and error occurs, returned task is cancelled.
         /// </summary>
-        public Task WaitForStateChangedAsync(ChannelState lastObservedState, DateTime? deadline = null)
+        public async Task WaitForStateChangedAsync(ChannelState lastObservedState, DateTime? deadline = null)
+        {
+            var result = await WaitForStateChangedInternalAsync(lastObservedState, deadline).ConfigureAwait(false);
+            if (!result)
+            {
+                throw new TaskCanceledException("Reached deadline.");
```

Comment:
Not certain here, but this causes the background watcher to loop to break out and finish as soon as one of the 1 second state-change-deadlines is reached, no?

Reply:
RunConnectivityWatcherAsync() now uses  WaitForStateChangedInternalAsync() instead (the "internal" version doesn't throw, which is the whole purpose of this change).